File: app/3-voice assistent + playback/vosk-android-demo/app/src/main/java/org/vosk/demo/VoskRecognitionService.java
package org.vosk.demo;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.Service;
import android.content.Intent;
import android.os.Build;
import android.os.IBinder;
import android.os.PowerManager;
import android.util.Log;

import androidx.core.app.NotificationCompat;

import org.json.JSONException;
import org.json.JSONObject;
import org.vosk.Model;
import org.vosk.Recognizer;
import org.vosk.android.RecognitionListener;
import org.vosk.android.SpeechService;
import org.vosk.android.StorageService;

import java.io.File;

public class VoskRecognitionService extends Service implements RecognitionListener {

    private static final String TAG = "VoskService";
    private static final int NOTIFICATION_ID = 1234;
    private static final String CHANNEL_ID = "vosk_channel";

    public static final String ACTION_START_RECOGNITION = "org.vosk.demo.ACTION_START";
    public static final String ACTION_STOP_RECOGNITION = "org.vosk.demo.ACTION_STOP";

    private Model model;
    private SpeechService speechService;
    private PowerManager.WakeLock wakeLock;
    private boolean isRunning = false;

    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    @Override
    public void onCreate() {
        super.onCreate();
        createNotificationChannel();
        acquireWakeLock();
        startForeground(NOTIFICATION_ID, createNotification());
        initModel();
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        if (intent != null && intent.getAction() != null) {
            switch (intent.getAction()) {
                case ACTION_START_RECOGNITION:
                    startListening();
                    break;
                case ACTION_STOP_RECOGNITION:
                    stopListening();
                    break;
            }
        }
        return START_STICKY;
    }

    private void initModel() {
        if (model != null) {
            startListening();
            return;
        }

        StorageService.unpack(this, "model-en-us", "model",
                (model) -> {
                    this.model = model;
                    startListening();
                },
                (exception) -> {
                    Log.e(TAG, "Model unpack failed", exception);
                    stopSelf();
                });
    }

    private void startListening() {
        if (isRunning || model == null) return;

        try {
            Recognizer recognizer = new Recognizer(model, 16000.0f);
            speechService = new SpeechService(recognizer, 16000.0f);
            speechService.startListening(this);
            isRunning = true;
            updateNotification();
        } catch (Exception e) {
            Log.e(TAG, "Error starting recognition", e);
        }
    }

    private void stopListening() {
        if (!isRunning) return;

        if (speechService != null) {
            speechService.stop();
            speechService.shutdown();
            speechService = null;
        }

        isRunning = false;
        updateNotification();
    }

    private void updateNotification() {
        NotificationManager manager = getSystemService(NotificationManager.class);
        manager.notify(NOTIFICATION_ID, createNotification());
    }

    private Notification createNotification() {
        return new NotificationCompat.Builder(this, CHANNEL_ID)
                .setContentTitle("Voice Recognition")
                .setContentText(isRunning ? "Listening..." : "Idle")
                .setSmallIcon(R.drawable.ic_mic)
                .setOngoing(true)
                .build();
    }

    private void createNotificationChannel() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            NotificationChannel channel = new NotificationChannel(
                    CHANNEL_ID,
                    "Vosk Voice Recognition",
                    NotificationManager.IMPORTANCE_LOW
            );
            getSystemService(NotificationManager.class).createNotificationChannel(channel);
        }
    }

    private void acquireWakeLock() {
        PowerManager pm = (PowerManager) getSystemService(POWER_SERVICE);
        wakeLock = pm.newWakeLock(PowerManager.PARTIAL_WAKE_LOCK, "VoskService::WakeLock");
        wakeLock.acquire();
    }

    @Override
    public void onDestroy() {
        stopListening();
        if (wakeLock != null && wakeLock.isHeld()) {
            wakeLock.release();
        }
        super.onDestroy();
    }

    @Override
    public void onResult(String hypothesis) {
        try {
            JSONObject json = new JSONObject(hypothesis);
            String text = json.getString("text").toLowerCase().trim();
            text = text.replaceAll("\\s+", " ");
            Log.d(TAG, "Recognized: " + text);

            String command = null;

            String[] greetSynonyms = {"hi", "hey", "high", "hello", "hail"};
            String[] carSynonyms = {"car", "call", "court", "assistant", "vehicle", "system", "core"};
            String[] lightSynonyms = {"light", "lite", "like", "lamp", "bulb", "glow"};
            String[] modeSynonyms = {"mode", "mood", "move", "moved", "setting", "style" , "moon"};
            String[] maxSynonyms = {"maximum", "max", "full"};
            String[] minSynonyms = {"minimum", "min", "meme", "low", "small"};
            String[] daySynonyms = {"day", "they", "the"};
            String[] nightSynonyms = {"night", "nine"};
            String[] seatSynonyms = {"seat", "seek", "see" , "the"};
            String[] offSynonyms = {"off", "of", "oh"};

            String greetPattern = String.join("|", greetSynonyms);
            String carPattern = String.join("|", carSynonyms);
            String lightPattern = String.join("|", lightSynonyms);
            String modePattern = String.join("|", modeSynonyms);
            String maxPattern = String.join("|", maxSynonyms);
            String minPattern = String.join("|", minSynonyms);
            String dayPattern = String.join("|", daySynonyms);
            String nightPattern = String.join("|", nightSynonyms);
            String seatPattern = String.join("|", seatSynonyms);
            String offPattern = String.join("|", offSynonyms);

            if (text.matches(".*\\b(" + greetPattern + ")\\s+(" + carPattern + ")\\b.*")) {
                command = "greet";
            }else if (text.matches(".*\\b(" + seatPattern + ")\\s+default\\b.*")) {
                command = "seat_70";
            } else if (text.matches(".*\\b(" + seatPattern + ")\\s+(" + maxPattern + ")\\b.*")) {
                command = "seat_90";
            }else if (text.matches(".*\\b(" + seatPattern + ")\\s+(" + minPattern + ")\\b.*")) {
                command = "seat_45";
            } else if (text.matches(".*\\b(" + lightPattern + ")\\s+on\\b.*")) {
                command = "light_on";
            } else if (text.matches(".*\\b(" + lightPattern + ")\\s+(" + offPattern + ")\\b.*")) {
                command = "light_off";
            } else if (text.matches(".*\\b(" + dayPattern + ")\\s+(" + modePattern + ")\\b.*")) {
                command = "day_mode";
            } else if (text.matches(".*\\b(" + nightPattern + ")\\s+(" + modePattern + ")\\b.*")) {
                command = "night_mode";
            }

            if (command != null) {
                Intent intent = new Intent("org.vosk.demo.ACTION_COMMAND");
                intent.putExtra("command", command);
                sendBroadcast(intent);
            } else {
                Log.d(TAG, "Unrecognized command: " + text);
            }

        } catch (JSONException e) {
            Log.e(TAG, "JSON parsing error", e);
        }
    }


    @Override public void onFinalResult(String hypothesis) {}
    @Override public void onPartialResult(String hypothesis) {}
    @Override public void onError(Exception e) { stopListening(); }
    @Override public void onTimeout() { stopListening(); }
}

File: App/DriverLauncher/app/src/main/java/com/example/driverlauncher/voskva/VoskRecognitionService.kt
package com.example.driverlauncher.voskva

import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.Service
import android.content.Intent
import android.os.Binder
import android.os.Build
import android.os.IBinder
import android.os.PowerManager
import android.util.Log
import androidx.core.app.NotificationCompat
import com.example.driverlauncher.R
import org.json.JSONException
import org.json.JSONObject
import org.vosk.Model
import org.vosk.Recognizer
import org.vosk.android.RecognitionListener
import org.vosk.android.SpeechService
import org.vosk.android.StorageService

class VoskRecognitionService : Service(), RecognitionListener {

    companion object {
        const val TAG = "VoskService"
        const val NOTIFICATION_ID = 1234
        const val CHANNEL_ID = "vosk_channel"
        const val ACTION_START_RECOGNITION = "org.vosk.demo.ACTION_START"
        const val ACTION_STOP_RECOGNITION = "org.vosk.demo.ACTION_STOP"
    }

    private var model: Model? = null
    private var speechService: SpeechService? = null
    private var wakeLock: PowerManager.WakeLock? = null
    private var isRunning = false
    private var callback: RecognitionCallback? = null

    private val binder = LocalBinder()

    inner class LocalBinder : Binder() {
        fun getService(): VoskRecognitionService = this@VoskRecognitionService
    }

    override fun onBind(intent: Intent): IBinder = binder

    fun setRecognitionCallback(callback: RecognitionCallback) {
        this.callback = callback
    }

    override fun onCreate() {
        super.onCreate()
        createNotificationChannel()
        acquireWakeLock()
        startForeground(NOTIFICATION_ID, createNotification())
        Log.i("TAG", "onCreate: Vosk Service Started With Notification")
        initModel()
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        when (intent?.action) {
            ACTION_START_RECOGNITION -> startListening()
            ACTION_STOP_RECOGNITION -> stopListening()
        }
        return START_STICKY
    }

    private fun initModel() {
        if (model != null) {
            startListening()
            return
        }

        StorageService.unpack(this, "vosk-model-small-en-us-0.15", "model",
            { unpackedModel ->
                model = unpackedModel
                startListening()
            },
            { exception ->
                Log.e(TAG, "Model unpack failed", exception)
                stopSelf()
            })

        Log.i("TAG", "initModel: Vosk Initialized Successfully")
    }

    private fun startListening() {
        if (isRunning || model == null) return

        try {
            val recognizer = Recognizer(model, 16000.0f)
            speechService = SpeechService(recognizer, 16000.0f)
            speechService?.startListening(this)
            isRunning = true
            updateNotification()
        } catch (e: Exception) {
            Log.e(TAG, "Error starting recognition", e)
        }
    }

    private fun stopListening() {
        if (!isRunning) return

        speechService?.apply {
            stop()
            shutdown()
        }
        speechService = null
        isRunning = false
        updateNotification()
    }

    private fun updateNotification() {
        getSystemService(NotificationManager::class.java)?.notify(NOTIFICATION_ID, createNotification())
    }

    private fun createNotification(): Notification {
        return NotificationCompat.Builder(this, CHANNEL_ID)
            .setContentTitle("Voice Recognition")
            .setContentText(if (isRunning) "Listening..." else "Idle")
            .setSmallIcon(R.drawable.voice)
            .setOngoing(true)
            .build()
    }

    private fun createNotificationChannel() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val channel = NotificationChannel(
                CHANNEL_ID,
                "Vosk Voice Recognition",
                NotificationManager.IMPORTANCE_LOW
            )
            getSystemService(NotificationManager::class.java)?.createNotificationChannel(channel)
        }
    }

    private fun acquireWakeLock() {
        val pm = getSystemService(POWER_SERVICE) as PowerManager
        wakeLock = pm.newWakeLock(PowerManager.PARTIAL_WAKE_LOCK, "VoskService::WakeLock")
        wakeLock?.acquire(10*60*1000L /*10 minutes*/)
    }

    override fun onDestroy() {
        stopListening()
        wakeLock?.takeIf { it.isHeld }?.release()
        super.onDestroy()
    }

    override fun onResult(hypothesis: String) {
        try {
            val json = JSONObject(hypothesis)
            var text = json.getString("text")
                .lowercase()
                .trim()
                .replace("\\s+".toRegex(), " ")

            Log.d(TAG, "Recognized: $text")

            val command = when {
                matchesPattern(text, greetSynonyms, carSynonyms) -> "greet"
                matchesPattern(text, seatSynonyms, "default") -> "seat_70"
                matchesPattern(text, seatSynonyms, maxSynonyms) -> "seat_90"
                matchesPattern(text, seatSynonyms, minSynonyms) -> "seat_45"
                matchesPattern(text, lightSynonyms, "on") -> "light_on"
                matchesPattern(text, lightSynonyms, offSynonyms) -> "light_off"
                matchesPattern(text, daySynonyms, modeSynonyms) -> "day_mode"
                matchesPattern(text, nightSynonyms, modeSynonyms) -> "night_mode"
                else -> null
            }

            command?.let {
                callback?.onCommandReceived(it)
            } ?: Log.d(TAG, "Unrecognized command: $text")

        } catch (e: JSONException) {
            Log.e(TAG, "JSON parsing error", e)
        }
    }

    private fun matchesPattern(text: String, group1: Array<String>, group2: Any): Boolean {
        val pattern1 = group1.joinToString("|")
        val pattern2 = when (group2) {
            is Array<*> -> (group2 as Array<String>).joinToString("|")
            is String -> group2
            else -> return false
        }
        return text.matches(".*\\b($pattern1)\\s+($pattern2)\\b.*".toRegex())
    }

    // Command synonym lists
    private val greetSynonyms = arrayOf("hi", "hey", "high", "hello", "hail")
    private val carSynonyms = arrayOf("car", "call", "court", "assistant", "vehicle", "system", "core")
    private val lightSynonyms = arrayOf("light", "lite", "like", "lamp", "bulb", "glow")
    private val modeSynonyms = arrayOf("mode", "mood", "move", "moved", "setting", "style", "moon")
    private val maxSynonyms = arrayOf("maximum", "max", "full")
    private val minSynonyms = arrayOf("minimum", "min", "meme", "low", "small")
    private val daySynonyms = arrayOf("day", "they", "the")
    private val nightSynonyms = arrayOf("night", "nine")
    private val seatSynonyms = arrayOf("seat", "seek", "see", "the")
    private val offSynonyms = arrayOf("off", "of", "oh")

    // RecognitionListener stubs
    override fun onFinalResult(hypothesis: String?) = Unit
    override fun onPartialResult(hypothesis: String?) = Unit
    override fun onError(exception: Exception?) = stopListening()
    override fun onTimeout() = stopListening()

    // Callback interface
    fun interface RecognitionCallback {
        fun onCommandReceived(command: String)
    }
}